New View: Server Stats

Vito0912 opened the pull request:
Based on #3330

This PR introduces a new page titled "All Libraries Stats," which mirrors the existing "Library Stats" view but combines data from all libraries.

We might want to consider refactoring the design since this results in some duplicated code. While there are slight modifications, primarily in the computed section, much of the functionality remains the same.

Image:
![image](https://github.com/user-attachments/assets/573755da-c7f9-4dde-b72e-2bf7768e8f21)

Vito0912 committed:
added libraries/stats endpoint

Vito0912 committed:
fixed wrong ide completions

Vito0912 committed:
added library to be returned instead of libraryId and fixed stupid req.library asignment

Vito0912 committed:
copy pasted and edited some parts in all-stats

Vito0912 committed:
refactored libraries-stats and changed path

Vito0912 committed:
removed redirect

faush01 commented:
cant this just be a toggle on the stats page to include all libraries?
having multiple pages for stats like this may be confusing.

Vito0912 committed:
Merge branch 'advplyr:master' into feat/all-stats-page

Vito0912 committed:
first draft controller

Vito0912 committed:
updated controller

Vito0912 committed:
delete old stats view and added all library stats view

Vito0912 committed:
fixed controller to include id of library without changing the schema

Vito0912 committed:
optimized all stats

Vito0912 committed:
added client code

Vito0912 committed:
Merge branch 'advplyr:master' into feat/all-stats-page

Vito0912 committed:
Merge branch 'feat/all-stats-page' of https://github.com/Vito0912/audiobookshelf into feat/all-stats-page

Vito0912 committed:
removed old artifacts

Vito0912 commented:
This is now an uniform view:
![grafik](https://github.com/user-attachments/assets/0eaf1650-3d72-4feb-9907-86653b035794)
![grafik](https://github.com/user-attachments/assets/930764f6-5b51-479a-9813-8739dd743d07)

Genres, Authors, Longest Items (hrs) and Largest Items are merged together and using the biggest values across

advplyr commented:
I'm not a fan of merging podcast library and audiobook library stats and using the same stat info.

What is the motivation for an all library stats page? What is the data you are actually interested in seeing across all libraries?

Vito0912 commented:
Are there specific reasons for this?

Generally, this overview should be an overview of all media on the server.
I've been planning to split my libraries for a while, but then I'd lose the statistics about total items, authors etc. broken down across everything.

In principle, I think it would be a quick task to differentiate between podcast and book media types, and then add up all podcasts for a podcast library and all books for book libraries. But I don't quite understand exactly why (Yes, podcasts are usually longer and larger), but that's basically irrelevant for the overall overview. Perhaps, and this would solve the problem, each episode could be considered as an item for the statistics.

But I am open to change that

nichwall commented:
I did some research about the CodeQL failure because that seems to be unrelated to this PR. For reference, the error is from `ApiRouter.js`, where the matching for paths is case sensitive. I believe this really only matters for the API Cache Manager and isn't really an error, but probably something good to handle globally.

I've found two main ways to fix it. Both are working in my local testing.

1. Convert paths to lowercase before matching
```
    // Middleware to convert request paths to lowercase
    const caseInsensitivePaths = (req, res, next) => {
      req.url = req.url.toLowerCase();
      next();
    };

    // Apply the middleware to all routes
    this.router.use(caseInsensitivePaths);

    // Define all of the existing routes
    this.router.get(/^\/libraries/, this.apiCacheManager.middleware)
    this.router.post('/libraries', LibraryController.create.bind(this))
    // rest of the routes...
```

2. Change the regex to be case insensitive.
```
this.router.get(/^\/libraries\/stats$/i, LibraryController.allStats.bind(this));
```

nichwall commented:
Also, responding to the design thing. I think having a view for all server stats together makes sense, but I agree it feels a bit weird to have stats from two different formats (books and podcasts) showing up together.

Maybe it makes sense for the "server stats" to show up on the Libraries page, like below all libraries? I don't think it really fits with this page, but also feel like it's too little information for its own "server stats page".
![Screenshot from 2025-01-01 17-10-14](https://github.com/user-attachments/assets/520b68a0-2270-4796-90e3-1b858e2932a3)

We could also make the server stats page more generic, and just show library summaries in a table instead of showing all of the "percent breakdown" for genres, size, top authors, etc. To view the "top authors" or "top genres" for a library, you could click on the library name and navigate to the stats page for that library.
![PXL_20250102_001650535 MP](https://github.com/user-attachments/assets/1b5fa367-6504-4751-818e-ce8e39ee1943)

Vito0912 committed:
updated view

Vito0912 committed:
fixed side bar, language and layout

Vito0912 committed:
added stats to podcast view

nichwall commented:
Stealing the image you posted on Discord for reference...
![image](https://github.com/user-attachments/assets/9507a9ab-352e-48cb-9359-a423bf2c380c)

Just tested it and it is working great for me. The only thing I noticed is that if you are on one podcast and click on a library that you are not currently in, the "library stats" page goes to the current library page instead of the library you clicked on. For example, regardless of if you click on "Books" or "Podcasts" in the above image, it will go to the "Audiobooks" library stats.

Vito0912 committed:
reverted old changes and fixed library switching

Vito0912 commented:
Waking up to an email starting with "Stealing the image you posted" ^^.

I fixed the clicking issue. Now it should switch and display the correct library. @advplyr I'm not sure if this is the proper way to do it. I just borrowed the code from the library switcher.

I also reverted the changes I made to the library stats view.

Vito0912 committed:
reverted old artifact agains

Vito0912 committed:
Merge branch 'advplyr:master' into feat/all-stats-page

Vito0912 commented:
@advplyr Before I resolve the merge conflicts and test it with the current branch I want to ask if this is still in the range of being merged or if you think this should be closed

advplyr commented:
I was thinking about this and set up a new stats controller in #4168. I'll look at it some more tomorrow

Vito0912 commented:
Hey, I wanted to ask if you had time to take a look at this already ?

Vito0912 committed:
Merge branch 'master' into feat/all-stats-page

Vito0912 committed:
fix podcast
